rosbag2_compression/src/rosbag2_compression/zstd_compressor.cpp

```diff
@@ -0,0 +1,120 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <memory>
+#include <string>
+
+#include "logging.hpp"
+#include "rosbag2_compression/zstd_compressor.hpp"
+
+namespace
+{
+// Increasing the compression level will:
+//   - Increase the time taken to compress
+//   - Decrease the size of the compressed data
+// Setting to zero uses Zstd's default value of 3.
+constexpr const int DEFAULT_ZSTD_COMPRESSION_LEVEL = 1;
+
+std::unique_ptr<char[]> get_input_buffer(
+  const std::string & uri,
+  size_t & decompressed_buffer_length)
+{
+  try {
+    std::ifstream infile{uri};
+    infile.exceptions(std::ifstream::failbit);
+    // Get size and allocate
+    infile.seekg(0, std::ios::end);
+    decompressed_buffer_length = infile.tellg();
+    auto decompressed_buffer = std::make_unique<char[]>(decompressed_buffer_length);
+    // Go back and read in contents
+    infile.seekg(0 /* off */, std::ios::beg);
+    infile.read(decompressed_buffer.get(), decompressed_buffer_length);
+    return decompressed_buffer;
+  } catch (std::ios_base::failure & fail) {
+    ROSBAG2_COMPRESSION_LOG_ERROR_STREAM(
+      "Caught IO stream exception while reading file: " << fail.what());
+    throw fail;
+  }
+}
+
+void check_compression_result(const size_t compression_result)
+{
+  if (ZSTD_isError(compression_result)) {
+    std::stringstream error;
+    error << "ZSTD compression error: " << ZSTD_getErrorName(compression_result);
+    throw std::runtime_error(error.str());
+  }
+  ROSBAG2_COMPRESSION_LOG_DEBUG("ZSTD compressed file.");
+}
+
+void print_compression_statistics(
+  std::chrono::high_resolution_clock::time_point start,
+  std::chrono::high_resolution_clock::time_point end,
+  size_t decompressed_size, size_t compressed_size)
```

Comment:
if both, `decompressed_size` and `compressed_size` are only used after being casted to `double`, why not receiving the parameters as such?

Reply:
Since that's their default type when received from Zstd/getting file size from the input file stream.